### PR TITLE
Exception proposal

### PR DESCRIPTION
**Pull request by SimonCropp**
fixes https://github.com/Azure/azure-service-bus-dotnet/issues/303

i know u said dont do a PR. but it is helpful to see the difference in code as part of the discussion. so while i dont necessarily expect this to be merged, this would be my proposed new state

 * 820 lines less code 
 * shaves 8K off the dll+pdb

### CODE REVIEW

**Comment by SeanFeldman** [comment]
:+1:

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 1]
```diff
@@ -61,7 +60,8 @@ public string MessageId
 
             set
             {
-                Message.ValidateMessageId(value);
+                Guard.AgainstNullAndEmpty(nameof(MessageId), value);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >AgainstNullAndEmpty [](start = 22, length = 19)

Lets change this. I can see scenarios where someone wants to clear the assigned `MessageId`. Lets just guard against null here

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 2]
```diff
@@ -76,7 +76,8 @@ public string PartitionKey
 
             set
             {
-                Message.ValidatePartitionKey(nameof(this.PartitionKey), value);
+                Guard.AgainstNullAndEmpty(nameof(PartitionKey), value);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >AgainstNullAndEmpty [](start = 22, length = 19)

Guard only against null

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 3]
```diff
@@ -90,7 +91,8 @@ public string ViaPartitionKey
 
             set
             {
-                Message.ValidatePartitionKey(nameof(this.ViaPartitionKey), value);
+                Guard.AgainstNullAndEmpty(nameof(ViaPartitionKey), value);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >AgainstNullAndEmpty [](start = 22, length = 19)

Guard only against null

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 4]
```diff
@@ -104,7 +106,8 @@ public string SessionId
 
             set
             {
-                Message.ValidateSessionId(nameof(this.SessionId), value);
+                Guard.AgainstNullAndEmpty(nameof(SessionId), value);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >AgainstNullAndEmpty [](start = 22, length = 19)

 Guard only against null

File: src/Microsoft.Azure.ServiceBus/Message.cs [hunk 5]
```diff
@@ -118,7 +121,8 @@ public string ReplyToSessionId
 
             set
             {
-                Message.ValidateSessionId(nameof(this.ReplyToSessionId), value);
+                Guard.AgainstNullAndEmpty(nameof(ReplyToSessionId), value);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Message.cs]: >AgainstNullAndEmpty [](start = 22, length = 19)

Guard only against null

File: src/Microsoft.Azure.ServiceBus/ServiceBusConnectionStringBuilder.cs [hunk 6]
```diff
@@ -59,14 +58,10 @@ public ServiceBusConnectionStringBuilder(string connectionString)
...
-                throw Fx.Exception.ArgumentNullOrWhiteSpace(nameof(endpoint));
-            }
-            if (string.IsNullOrWhiteSpace(sharedAccessKeyName) || string.IsNullOrWhiteSpace(sharedAccessKey))
-            {
-                throw Fx.Exception.ArgumentNullOrWhiteSpace(string.IsNullOrWhiteSpace(sharedAccessKeyName) ? nameof(sharedAccessKeyName) : nameof(sharedAccessKey));
-            }
+            Guard.AgainstNullAndEmpty(nameof(endpoint), endpoint);
+            Guard.AgainstNullAndEmpty(nameof(sharedAccessKeyName), sharedAccessKeyName);
+            Guard.AgainstNullAndEmpty(nameof(sharedAccessKey), sharedAccessKey);
+            Guard.AgainstNullAndEmpty(nameof(entityPath), entityPath);
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/ServiceBusConnectionStringBuilder.cs]: Remove this line. EntityPath can be null or empty

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 7]
```diff
@@ -726,13 +691,13 @@ public override void UnregisterPlugin(string serviceBusPluginName)
...
                 receivingAmqpLink.Session.SafeClose();
-                throw new ServiceBusException(true, Resources.SessionFilterMissing);
+                throw new ServiceBusException(true, "Session filter is missing on the link.");
             }
 
             if (string.IsNullOrWhiteSpace(tempSessionId))
             {
                 receivingAmqpLink.Session.SafeClose();
-                throw new ServiceBusException(true, Resources.AmqpFieldSessionId);
+                throw new ServiceBusException(true, "No session-id was specified for a session receiver.");
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: >"No session-id was specified for a session receiver." [](start = 52, length = 53)

Could you update this to mention 
"Please retry the operation".

File: src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs [hunk 8]
```diff
@@ -1152,12 +1117,9 @@ static void CloseRequestResponseSession(RequestResponseAmqpLink requestResponseA
                     {
                         if (error.Condition.Equals(AmqpErrorCode.NotFound))
                         {
-                            if (this.isSessionReceiver)
-                            {
-                                throw new SessionLockLostException(Resources.SessionLockExpiredOnMessageSession);
-                            }
+                            ThrowIfSessionReceiver();
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs]: >ThrowIfSessionReceiver [](start = 28, length = 22)

This name sounds misleading. Rename to something like
`ThrowSessionLostIfSessionReceiver`

File: src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs [hunk 9]
```diff
@@ -397,18 +368,15 @@ static void CloseRequestResponseSession(RequestResponseAmqpLink requestResponseA
                         var size = (ulong)amqpMessage.SerializedMessageSize;
                         if (size > amqpLink.Settings.MaxMessageSize.Value)
                         {
-                            // TODO: Add MessageSizeExceededException
-                            throw new NotImplementedException("MessageSizeExceededException: " + Resources.AmqpMessageSizeExceeded.FormatForUser(amqpMessage.DeliveryId.Value, size, amqpLink.Settings.MaxMessageSize.Value));
-                            ////throw Fx.Exception.AsError(new MessageSizeExceededException(
-                            ////Resources.AmqpMessageSizeExceeded.FormatForUser(amqpMessage.DeliveryId.Value, size, amqpLink.Settings.MaxMessageSize.Value)));
+                            throw new NotImplementedException($"The received message (delivery-id:{amqpMessage.DeliveryId.Value}, size:{size} bytes) exceeds the limit ({amqpLink.Settings.MaxMessageSize.Value} bytes) currently allowed on the link.");
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs]: >throw new NotImplementedException($"The received message [](start = 28, length = 56)

Retain the `MessageSizeExceededException` word in the exception message.

**Comment by nemakam** [comment]
@SimonCropp  Also, check the appveyor build

**Comment by SimonCropp** [comment]
@nemakam before i address individual feedback, is this an approach u want to go with. Ie no localisation of exception messages?

**Comment by nemakam** [comment]
@SimonCropp, yes. Lets go ahead with this approach

**Comment by SeanFeldman** [comment]
@nemakam, so `on hold` in #303 is just for decoration? :joy:

**Comment by nemakam** [comment]
Done! Its a go!

**Comment by clemensv** [comment]
@SimonCropp Look like @nemakam cleared this after fixing the remaining blockers?

**Comment by SimonCropp** [comment]
Given u wont be inlining resources this pr may as well be abandoned and a new one started

**Pull request closed without merging**